spec.html

```diff
@@ -35994,9 +36000,7 @@ <h1>
                 1. Assert: _r_ is a State.
                 1. Set _matchSucceeded_ to *true*.
             1. Let _e_ be _r_'s _endIndex_ value.
-            1. If _fullUnicode_ is *true*, then
-              1. _e_ is an index into the _Input_ character list, derived from _S_, matched by _matcher_. Let _eUTF_ be the smallest index into _S_ that corresponds to the character at element _e_ of _Input_. If _e_ is greater than or equal to the number of elements in _Input_, then _eUTF_ is the number of code units in _S_.
-              1. Set _e_ to _eUTF_.
+            1. If _fullUnicode_ is *true*, set _e_ to ! GetStringIndex(_S_, _Input_, _e_).
```

Comment:
I don’t think Inout is defined anywhere now - and it seems like it should be derived from S inside StringGetIndex rather than passing it in?

Reply:
When was that changed?

Reply:
Isn't it defined in https://tc39.es/ecma262/#sec-pattern-semantics?

Reply:
I think it’s defined there for the *other* places you use it, but i think this one doesn’t use that, which is why the existing prose defines it inline.

Reply:
Would it be sufficient to change the algorithm steps of GetStringIndex to:

```
1. If _S_ is the empty String, return 0.
1. Let _Input_ be StringToCodePoints(_S_).
1. Let _eUTF_ be the smallest index into _S_ that corresponds to the character at element _e_ of _input_. If _e_ is greater than or equal to the number of elements in _input_, then _eUTF_ is the number of code units in _S_.
1. Return _eUTF_.
```

and let implementations optimize away the repeated invocations of `StringToCodePoints(_S_)`? It feels very hand-wavy (then again, so did the previous wording this replaces).

Something less hand-wavy might have created a List of code points and a List of the original index in `_S_` at which each code point was found, and passed them both to `GetStringIndex` (or a single list containing tuples of `(codePoint, index)`).

Reply:
Or should `GetStringIndex` just contain a loop that repeated calls `AdvanceStringIndex` until enough code points have been consumed?

Reply:
i mean, there's also `StringIndexOf` already, so it's possible nothing new is needed here.

In particular it just seems silly to pass around `S` and `Input` when both are, via assertion, always matching and derivable from each other, and when the spec isn't a programming language so implementations can trivially optimize it (meaning, we shouldn't be worried about the perf of the spec algorithm, only about the perf of resulting implementations)

Reply:
> I think it’s defined there for the other places you use it, but i think this one doesn’t use that, which is why the existing prose defines it inline.

The original prose doesn't explain how _Input_ is derived from _S_, only that it **is** derived from _S_. https://tc39.es/ecma262/#sec-pattern-semantics is the only place that describes how _Input_ is to be derived.

Reply:
_e_ refers to the position in _Input_ per the pattern semantics. The point of this operation is to convert the ordinal index of _e_ in _Input_ to an actual string index in _S_.

Reply:
Part of the problem with this section is that _Input_ is defined in https://tc39.es/ecma262/#sec-pattern-semantics, and initialized in https://tc39.es/ecma262/#sec-compilepattern, step 2.c, and is "used throughout the algorithms in 22.2.2", but isn't defined as an output of the matcher created by CompilePattern, so RegExpBuiltinExec has always been referencing something not exactly "reachable" by the algorithm.

Reply:
> i mean, there's also `StringIndexOf` already, so it's possible nothing new is needed here.

StringIndexOf doesn't do anything like what this is trying to do.

Reply:
The simplest thing may be just to revert this line, duplicate the original steps for _captureStart_ and _captureEnd_, and drop GetStringIndex altogether. We could then rework RegExpBuiltinExec and the Regular expression pattern semantics later.

Reply:
I'm not a fan of how wordy it is, but the change needed to drop GetStringIndex would look something like this: https://github.com/tc39/ecma262/pull/1713#pullrequestreview-853492024

Reply:
Alternatively, the steps that define _Input_ and _listIndex_ in CompilePattern should be pulled out and placed in RegExpBuiltinExec instead, and passed as arguments to _matcher_ instead of _S_ and _lastIndex_, so that it can be referred to correctly.

Reply:
> Alternatively, the steps that define _Input_ and _listIndex_ in CompilePattern should be pulled out and placed in RegExpBuiltinExec instead, and passed as arguments to _matcher_ instead of _S_ and _lastIndex_, so that it can be referred to correctly.

At first glance, this seems to be a cleaner approach and removes some of the ambiguity around _Input_. I'll amend this PR with this change and we can discuss it further.